app/services/resource_service.rb

```diff
@@ -15,101 +15,159 @@ def create(resource_class, params)
+    @external_resource = find_external_resource
+    @old_resource = find_old_resource
+    if @external_resource.present? && unchanged_attributes? && !always_recreate?
+      @external_resource.touch
+      return @old_resource
     end
 
-    create_or_recreate(old_record, external_resource)
-    resource_or_error_message(resource)
+    resource_or_error_message(update_or_create)
   end
 
-  def create_or_recreate(old_record, external_resource)
-    ActiveRecord::Base.transaction do
-      old_record.destroy if old_record.present?
-      external_resource.destroy if external_resource.present?
-      if @resource.save
-        create_external_resource
-        FacebookService.delay.send_post_to_page(resource_id: @resource.id, resource_type: @resource_class.name)
-        set_default_categories if @resource.respond_to?(:categories)
+  private
+
+    # Find old resource
+    #
+    # @return [Object] Instance of @resource_class
+    def find_old_resource
+      if @old_resource_id.present?
+        @resource_class.filtered_for_current_user(@data_provider.user).find_by(id: @old_resource_id)
       else
-        GraphQL::ExecutionError.new("Invalid input: #{@resource.errors.full_messages.join(', ')}")
+        @external_resource.try(:external)
       end
     end
-  end
 
-  def unchanged_attributes?(record)
-    return false if record.blank?
-    return @resource.compareable_attributes == record.compareable_attributes if record.respond_to?(:compareable_attributes)
+    def update_or_create
+      if @old_resource.present?
+        update_resource
+      else
+        create_new_resource
+      end
+    end
 
-    # Fallback, wenn :compareable_attributes nicht beim Model definiert ist.
-    except_attributes = ["id", "created_at", "updated_at", "tag_list", "category_id", "region_id", "visible"]
-    @resource.attributes.except(*except_attributes) == record.attributes.except(*except_attributes)
-  end
+    def update_resource
+      # find all association names to delete for a resource
+      association_names_to_delete = @resource_class
+                                      .reflect_on_all_associations
+                                      .select { |a| a.options[:dependent] == :destroy }.map(&:name)
+
+      # delete all nested resources
+      association_names_to_delete.each do |association_name|
+        associations = @old_resource.send(association_name)
+        associations.destroy_all if !singular?(association_name) && associations.any?
+        associations.destroy if singular?(association_name) && associations.present?
+      end
 
-  def always_recreate?(data_provider, resource_class)
-    return false if data_provider.blank?
+      # update all attributes and recreate nested resources
+      @old_resource.update(@params)
 
-    class_name = resource_class.name
-    setting = data_provider.data_resource_settings.where(data_resource_type: class_name).first
-    return false if setting.blank?
+      ExternalReference.create(
```

Comment:
Weil `external_reference` ein `dependent: :destroy` hat, wurde der alte Eintrag bereits gelöscht mit `association_names_to_delete`.

Reply:
Kommentar ergänzt in 0215c59